pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/customobject/DistinctTable.java

```diff
@@ -0,0 +1,297 @@
+      for (int i = 0; i < numColumns; i++) {
+        columnIndexMap.put(columns[i], i);
+      }
+      int numOrderByColumns = orderBy.size();
+      int[] orderByColumnIndexes = new int[numOrderByColumns];
+      boolean[] orderByAsc = new boolean[numOrderByColumns];
+      for (int i = 0; i < numOrderByColumns; i++) {
+        SelectionSort selectionSort = orderBy.get(i);
+        orderByColumnIndexes[i] = columnIndexMap.getInt(selectionSort.getColumn());
+        orderByAsc[i] = selectionSort.isIsAsc();
+      }
+      _sortedRecords = new PriorityQueue<>(initialCapacity, (record1, record2) -> {
+        Object[] values1 = record1.getValues();
+        Object[] values2 = record2.getValues();
+        for (int i = 0; i < numOrderByColumns; i++) {
+          Comparable valueToCompare1 = (Comparable) values1[orderByColumnIndexes[i]];
+          Comparable valueToCompare2 = (Comparable) values2[orderByColumnIndexes[i]];
+          int result =
+              orderByAsc[i] ? valueToCompare2.compareTo(valueToCompare1) : valueToCompare1.compareTo(valueToCompare2);
+          if (result != 0) {
+            return result;
+          }
+        }
+        return 0;
+      });
+    } else {
+      _sortedRecords = null;
+    }
+    _records = null;
+  }
+
+  /**
+   * Returns the {@code DataSchema} of the {@code DistinctTable}.
+   */
+  public DataSchema getDataSchema() {
+    return _dataSchema;
+  }
+
+  /**
+   * Returns the number of unique records within the {@code DistinctTable}.
+   */
+  public int size() {
+    if (_uniqueRecords != null) {
+      // Server-side
+      return _uniqueRecords.size();
+    } else {
+      // Broker-side
+      return _records.size();
+    }
+  }
+
+  /**
+   * Adds a record into the DistinctTable and returns whether more records should be added. No more records should be
+   * added iff there is no order-by column and enough unique records have been collected.
+   */
+  public boolean add(Record record) {
+    if (_uniqueRecords.contains(record)) {
+      return true;
+    }
+    if (_sortedRecords != null) {
+      if (_sortedRecords.size() < _limit) {
+        _uniqueRecords.add(record);
+        _sortedRecords.offer(record);
+      } else {
+        Record leastRecord = _sortedRecords.peek();
+        if (_sortedRecords.comparator().compare(record, leastRecord) > 0) {
+          _uniqueRecords.remove(leastRecord);
+          _uniqueRecords.add(record);
+          _sortedRecords.poll();
+          _sortedRecords.offer(record);
+        }
+      }
+      return true;
+    } else {
+      _uniqueRecords.add(record);
+      return _uniqueRecords.size() < _limit;
+    }
+  }
+
+  /**
+   * Returns {@code true} if no more records should be added, {@code false otherwise}. No more records should be added
+   * iff there is no order-by columns and enough unique records have been collected.
+   */
+  public boolean shouldNotAddMore() {
```

Comment:
So this can be useful for early termination for DISTINCT queries without ORDER BY?

Reply:
It can save the creation of `RowBasedBlockValueFetcher` within the `DistinctAggregationFunction.aggregate()`